### PR TITLE
Jackson raw value regression (#10505)

### LINKED ISSUE #10505: Problem serializing to JSON when upgrading from 2.8.2 to 2.8.3
### Play Version

2.8.3 

### API

Java API

### Operating System

Linux NBSL1380 5.4.0-52-generic #57-Ubuntu SMP Thu Oct 15 10:57:00 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux

### JDK

openjdk version "11.0.8" 2020-07-14
OpenJDK Runtime Environment AdoptOpenJDK (build 11.0.8+10)
OpenJDK 64-Bit Server VM AdoptOpenJDK (build 11.0.8+10, mixed mode)

### Expected Behavior

`play.libs.Json.toJson` is able to serialise a Jackson serializable object as it was in 2.8.2

### Actual Behavior

Our tests now fail with stack trace provided below: 
```
      java.lang.RuntimeException: java.lang.RuntimeException: java.lang.RuntimeException: We should have been reading an object, something got wrong
        at controllers.DecisionStrategyController.lambda$getDecisionStrategy$0(DecisionStrategyController.java:50) Test / executeTests 19s
	at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:642)
	at java.base/java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:478)
	at play.core.j.HttpExecutionContext.$anonfun$execute$1(HttpExecutionContext.scala:64)
	at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:48)
	at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:48)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
	at ✽.I should have a strategy with proxy settings attached(file:///home/tom/deko-docker/src/decisioning/test/cucumber/features/ProxySettings.feature:13)
Caused by: java.lang.RuntimeException: java.lang.RuntimeException: We should have been reading an object, something got wrong
	at play.libs.Json.toJson(Json.java:94)
	at controllers.DecisionStrategyController.lambda$getDecisionStrategy$0(DecisionStrategyController.java:48)
	at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:642)
	at java.base/java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:478)
	at play.core.j.HttpExecutionContext.$anonfun$execute$1(HttpExecutionContext.scala:64)
	at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:48)
	at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:48)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
Caused by: java.lang.RuntimeException: We should have been reading an object, something got wrong
	at play.utils.JsonNodeDeserializer.deserialize(JsonNodeDeserializer.scala:166)
	at play.utils.JsonNodeDeserializer.deserialize(JsonNodeDeserializer.scala:57)
	at play.utils.JsonNodeDeserializer.deserialize(JsonNodeDeserializer.scala:53)
	at com.fasterxml.jackson.databind.ObjectMapper._readValue(ObjectMapper.java:4497)
	at com.fasterxml.jackson.databind.ObjectMapper.readTree(ObjectMapper.java:2840)
	at com.fasterxml.jackson.databind.ObjectMapper.valueToTree(ObjectMapper.java:3229)
	at play.libs.Json.toJson(Json.java:92)
	... 11 more
```

Maybe a downstream issue with https://github.com/playframework/play-json?

### Reproducible Test Case

I'm unable to provide a reproducible test case at this time, I will try and spend some time later if possible and work out where the issue is.

### PR DESCRIPTION
**Pull request by ignasi35**
Fixes #10505

### CODE REVIEW

**Comment by ignasi35** [comment]
@mergifyio backport master

File: core/play/src/main/scala/play/utils/JsonNodeDeserializer.scala [hunk 1]
```diff
@@ -111,6 +111,25 @@ private class JsonNodeDeserializer extends JsonDeserializer[JsonNode] {
...
+    // should this occur?
+    if (ob == null) return nodeFactory.nullNode
+
+    val `type` = ob.getClass
+    if (`type` eq classOf[Array[Byte]]) return nodeFactory.binaryNode(ob.asInstanceOf[Array[Byte]]) // most common special case
+    if (ob.isInstanceOf[RawValue]) return nodeFactory.rawValueNode(ob.asInstanceOf[RawValue])
+    if (ob.isInstanceOf[JsonNode]) return ob.asInstanceOf[JsonNode]
+
+    nodeFactory.pojoNode(ob)
+  }
```
Comment (ignasi35) [inline core/play/src/main/scala/play/utils/JsonNodeDeserializer.scala:129]: taken (almost) verbatim from https://github.com/FasterXML/jackson-databind/blob/57482db32f4490f17615e0540a9fdf1c7784abb0/src/main/java/com/fasterxml/jackson/databind/deser/std/JsonNodeDeserializer.java#L679-L700

File: core/play/src/main/scala/play/utils/JsonNodeDeserializer.scala [hunk 2]
```diff
@@ -160,10 +179,9 @@ private class JsonNodeDeserializer extends JsonDeserializer[JsonNode] {
         }
 
       case JsonTokenId.ID_NOT_AVAILABLE =>
-        throw new RuntimeException("We should have been reading an object, something got wrong")
+        throw new RuntimeException("Didn't receive a token when requesting one. See Jackson's JsonToken#NOT_AVAILABLE.")
```
Comment (ignasi35) [inline core/play/src/main/scala/play/utils/JsonNodeDeserializer.scala:182]: Improved this exception message as I was passing by.

File: core/play-integration-test/src/it/scala/play/it/libs/json/JavaJsonSpec.scala [hunk 3]
```diff
@@ -110,6 +110,23 @@ trait JavaJsonSpec extends Specification {
...
+            | "city": "NYC",
+            | "street": "Manhattan Av."
+            |}
+            |""".stripMargin
+
+        private val jsonNode: JsonNode = Json.parse(json)
+        private val customer: Customer = Json.fromJson(jsonNode, classOf[Customer])
+        customer.code must_== "1234"
+        customer.address must_== new Address("NYC", "Manhattan Av.")
+      }
```
Comment (ignasi35) [inline core/play-integration-test/src/it/scala/play/it/libs/json/JavaJsonSpec.scala:128]: Wrapped objects worked as expected already. I only stumbled upon this Jackson feature and added the tests to be double-sure.

File: core/play-integration-test/src/it/scala/play/it/libs/json/JavaJsonSpec.scala [hunk 4]
```diff
@@ -141,6 +158,43 @@ trait JavaJsonSpec extends Specification {
...
+
+      "unwrapped objects" in new JsonScope {
+        val c = new Customer("1234")
+        c.address = new Address("NYC", "Manhattan Av.")
+        val json = Json.stringify(Json.toJson(c))
+        json must_== """{"code":"1234","city":"NYC","street":"Manhattan Av."}"""
+
+      }
+
+      "embedded/raw object" in new JsonScope {
```
Comment (ignasi35) [inline core/play-integration-test/src/it/scala/play/it/libs/json/JavaJsonSpec.scala:170]: Using `@JsonRawValue` only applies when serializing so there's no need to a test parsing a JSON string into `ShoppingCart`.

**Review by ignasi35 — COMMENTED** [review]
comments for reviewers

File: core/play/src/main/scala/play/utils/JsonNodeDeserializer.scala [hunk 5]
```diff
@@ -111,6 +111,25 @@ private class JsonNodeDeserializer extends JsonDeserializer[JsonNode] {
...
+    import com.fasterxml.jackson.databind.util.RawValue
+    val ob = p.getEmbeddedObject
+
+    // should this occur?
+    if (ob == null) return nodeFactory.nullNode
+
+    val `type` = ob.getClass
+    if (`type` eq classOf[Array[Byte]]) return nodeFactory.binaryNode(ob.asInstanceOf[Array[Byte]]) // most common special case
+    if (ob.isInstanceOf[RawValue]) return nodeFactory.rawValueNode(ob.asInstanceOf[RawValue])
+    if (ob.isInstanceOf[JsonNode]) return ob.asInstanceOf[JsonNode]
```
Comment (octonato) [inline core/play/src/main/scala/play/utils/JsonNodeDeserializer.scala:126]: we should avoid usage of `return` keyword

**Review by octonato — APPROVED** [review]
LGTM after removing the usage of `return`

**Comment by octonato** [comment]
@Mergifyio backport 2.7.x

**Pull request merged**